tst/DtlsFunctionalityTest.cpp

```diff
@@ -0,0 +1,143 @@
+
+        DtlsSessionOutboundPacketFunc outboundPacketFn = [](UINT64 customData, PBYTE pData, UINT32 dataLen) {
+            // Since dtlsSessionProcessPacket might send packets trhough dtlsSessionPutApplicationData which eventually calls this callback again,
+            // we need to make sure that the stack is not going to be too deep by creating another thread to detach from the main stack
+            std::thread([customData, pData, dataLen] {
+                PDtlsSession pPeer = (PDtlsSession) customData;
+                assert(pPeer != NULL);
+                assert(pData != NULL);
+                CHK_LOG_ERR(dtlsSessionProcessPacket(pPeer, pData, (PINT32) &dataLen));
+            }).detach();
+        };
+
+        CHK_STATUS(createDtlsSession(&callbacks, timerQueueHandle, 0, FALSE, NULL, &pServer));
+        CHK_STATUS(createDtlsSession(&callbacks, timerQueueHandle, 0, FALSE, NULL, &pClient));
+
+        CHK_STATUS(dtlsSessionOnOutBoundData(pServer, (UINT64) pClient, outboundPacketFn));
+        CHK_STATUS(dtlsSessionOnOutBoundData(pClient, (UINT64) pServer, outboundPacketFn));
+
+        CHK_STATUS(dtlsSessionStart(pServer, FALSE));
+        CHK_STATUS(dtlsSessionStart(pClient, TRUE));
+
+        for (int i = 0; i < 100 && ATOMIC_LOAD(&connectedCount) != 2; i++) {
+            THREAD_SLEEP(20 * HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
+        }
+
+        CHK_ERR(ATOMIC_LOAD(&connectedCount) == 2, STATUS_OPERATION_TIMED_OUT, "timeout: failed to finish initial handshake");
+
+        *ppClient = pClient;
+        *ppServer = pServer;
+
+    CleanUp:
+
+        if (STATUS_FAILED(retStatus)) {
+            if (pClient != NULL) {
+                freeDtlsSession(&pClient);
+            }
+
+            if (pServer != NULL) {
+                freeDtlsSession(&pServer);
+            }
+        }
+
+        return STATUS_SUCCESS;
+    }
+};
+
+VOID outboundPacketFnNoop(UINT64 customData, PBYTE pData, UINT32 dataLen)
+{
+    UNUSED_PARAM(customData);
+    UNUSED_PARAM(pData);
+    UNUSED_PARAM(dataLen);
+}
+
+TEST_F(DtlsFunctionalityTest, putApplicationDataWithVariedSizes)
+{
+    PDtlsSession pClient = NULL, pServer = NULL;
+    TIMER_QUEUE_HANDLE timerQueueHandle = INVALID_TIMER_QUEUE_HANDLE_VALUE;
+    PBYTE pData = NULL;
+    INT32 dataSizes[] = {
+        4,     // very small packet
+        1000,  // small packet but should be still under mtu
+        10000, // big packet and bigger than even a jumbo frame
+    };
+
+    EXPECT_EQ(STATUS_SUCCESS, timerQueueCreate(&timerQueueHandle));
+    EXPECT_EQ(STATUS_SUCCESS, createAndConnect(timerQueueHandle, &pClient, &pServer));
+
+    EXPECT_EQ(STATUS_SUCCESS, dtlsSessionOnOutBoundData(pClient, 0, outboundPacketFnNoop));
+    EXPECT_EQ(STATUS_SUCCESS, dtlsSessionOnOutBoundData(pServer, 0, outboundPacketFnNoop));
+
+    for (int i = 0; i < (INT32) ARRAY_SIZE(dataSizes); i++) {
+        pData = (PBYTE) MEMREALLOC(pData, dataSizes[i]);
+        EXPECT_TRUE(pData != NULL);
```

Comment:
ASSERT_TRUE is better suited. Same for the rest of the tests below in the same cases

Reply:
Yep, forgot that google test is available here.

Reply:
Yep, makes sense